python/cuml/manifold/t_sne.pyx

```diff
@@ -0,0 +1,472 @@
+            self.perplexity = n
+
+        # Prepare output embeddings
+        Y = cuda.device_array(
+            (n, self.n_components),
+            order="F",
+            dtype=np.float32)
+
+        cdef uintptr_t embed_ptr = Y.device_ctypes_pointer.value
+
+        # Find best params if learning rate method is adaptive
+        if self.learning_rate_method=='adaptive' and self.method=="barnes_hut":
+            if self.verbose:
+                print("Learning rate is adpative. In TSNE paper, "
+                      "it has been shown that as n->inf, "
+                      "Barnes Hut works well if n_neighbors->30, "
+                      "learning_rate->20000, early_exaggeration->24.")
+                print("cuML uses an adpative method."
+                      "n_neighbors decreases to 30 as n->inf. "
+                      "Likewise for the other params.")
+            if n <= 2000:
+                self.n_neighbors = min(max(self.n_neighbors, 90), n)
+            else:
+                # A linear trend from (n=2000, neigh=100) to (n=60000,neigh=30)
+                self.n_neighbors = max(int(102 - 0.0012 * n), 30)
+            self.pre_learning_rate = max(n / 3.0, 1)
+            self.post_learning_rate = self.pre_learning_rate
+            self.early_exaggeration = 24.0 if n > 10000 else 12.0
+            if self.verbose:
+                print("New n_neighbors = {}, "
+                      "learning_rate = {}, "
+                      "exaggeration = {}".format(self.n_neighbors,
+                                                 self.pre_learning_rate,
+                                                 self.early_exaggeration))
+
+        assert(<void*> X_ptr != NULL and <void*> embed_ptr != NULL)
+
+        cdef long long seed = -1
+        if self.random_state is not None:
+            seed = self.random_state
+
+        TSNE_fit(handle_[0],
+                 <float*> X_ptr,
+                 <float*> embed_ptr,
+                 <int> n,
+                 <int> p,
+                 <int> self.n_components,
+                 <int> self.n_neighbors,
+                 <float> self.angle,
+                 <float> self.epssq,
+                 <float> self.perplexity,
+                 <int> self.perplexity_max_iter,
+                 <float> self.perplexity_tol,
+                 <float> self.early_exaggeration,
+                 <int> self.exaggeration_iter,
+                 <float> self.min_gain,
+                 <float> self.pre_learning_rate,
+                 <float> self.post_learning_rate,
+                 <int> self.n_iter,
+                 <float> self.min_grad_norm,
+                 <float> self.pre_momentum,
+                 <float> self.post_momentum,
+                 <long long> seed,
+                 <bool> self.verbose,
+                 <bool> True,
+                 <bool> True)
+
+        # Clean up memory
+        del _X
+        self.Y = Y.copy_to_host()
```

Comment:
This should not be getting copied to host. If this is being done for pickle serialization, see how we're handling this in other Cython classes (eg UMAP).

Reply:
Oh lol yep I forgot about this [this was to solve the strange illegal mem access issues]

Reply:
Corey is right, this should be left as a device array and then in the __getstate__ transform it to a cudf.Series so that it can be pickled. That said, this behavior will be added to Base pretty soon (today/tomorrow I’m submitting the PR), so you’ll inherit it automatically

Reply:
Cool! :) I fixed it now :)